File: Data Structures/#11.B-Trees & Red-Black Trees Exercises/Java/src/Hierarchy.java
import java.util.*;

public class Hierarchy<T> implements IHierarchy<T> {

    private Integer count;
    private Node<T> root;

    public Hierarchy(T element) {
        this.root = new Node<>(element);
        this.count = 1;
    }

    public void Add(T parent, T child) {

        Node<T> fatherResult = this.add(this.root, parent, child);

        if (fatherResult == null) {
            throw new IllegalArgumentException("Parent doesn't exists!");
        }

        Node<T> newNode = new Node<>(child, fatherResult);
        fatherResult.add(newNode);
        this.count++;
    }

    //Will search for the father and too if the new childValue exists already
    private Node<T> add(Node<T> start, T parentValue, T childValue) {

        Node<T> father = null;
        Queue<Node<T>> nodes = new ArrayDeque<>();
        nodes.add(start);

        while (nodes.size() != 0) {
            Node node = nodes.peek();

            if (node.getValue().equals(childValue)) {
                throw new IllegalArgumentException("Value already exists!");
            }

            if (node.getValue().equals(parentValue)) {
                father = node;
                break;
            }

            node.getChildrens().forEach(x -> {
                nodes.add((Node<T>) x);
            });

            nodes.poll();
        }

        return father;
    }

    public int getCount() {
        return this.count;
    }

    public void Remove(T element) {

        if (this.root.getValue().equals(element)) {
            throw new IllegalStateException("Collection is empty!");
        }

        int oldCount = this.count;
        Queue<Node<T>> nodes = new ArrayDeque<>();
        nodes.add(this.root);

        while (nodes.size() != 0) {
            Node node = nodes.peek();

            if (node.getValue().equals(element)) {
                Node<T> nodeFather = node.getFather();

                Set<Node<T>> newChildrens = new LinkedHashSet<>();

                nodeFather.getChildrens().remove(node);
                for (Object children : node.getChildrens()) {
                    Node<T> child = (Node<T>) children;
                    child.setFather(nodeFather);
                    nodeFather.add(child);
                }

                this.count--;
                break;
            }

            node.getChildrens().forEach(x -> {
                nodes.add((Node<T>) x);
            });

            nodes.poll();
        }

        if (oldCount == this.count) {
            throw new IllegalArgumentException();
        }

    }

    public boolean Contains(T element) {
        boolean isPresent = false;
        Queue<Node<T>> nodes = new ArrayDeque<>();
        nodes.add(this.root);

        while (nodes.size() != 0) {
            Node node = nodes.peek();

            if (node.getValue().equals(element)) {
                isPresent = true;
                break;
            }

            node.getChildrens().forEach(x -> {
                nodes.add((Node<T>) x);
            });

            nodes.poll();
        }

        return isPresent;
    }

    public T GetParent(T element) {

        if (this.count == 1 && this.root.getValue().equals(element)) {
            return null;
        }

        T fatherValue = null;
        Queue<Node<T>> nodes = new ArrayDeque<>();
        nodes.add(this.root);

        while (nodes.size() != 0) {
            Node node = nodes.peek();

            if (node.getValue().equals(element)) {
                Node<T> nodeFather = node.getFather();
                return nodeFather.getValue();
            }

            node.getChildrens().forEach(x -> {
                nodes.add((Node<T>) x);
            });

            nodes.poll();
        }

        throw new IllegalArgumentException("Not existing element");
    }

    public Iterable<T> GetChildren(T element) {

        Queue<Node<T>> nodes = new ArrayDeque<>();
        nodes.add(this.root);

        while (nodes.size() != 0) {
            Node node = nodes.peek();

            if (node.getValue().equals(element)) {
                List<T> values = new ArrayList<>();

                for (Object item : node.getChildrens()) {
                    Node<T> itemm = (Node<T>) item;
                    values.add(itemm.getValue());
                }

                return () -> values.iterator();
            }

            node.getChildrens().forEach(x -> {
                nodes.add((Node<T>) x);
            });

            nodes.poll();
        }

        throw new IllegalArgumentException("Not existing element!");
    }

    public Iterable<T> GetCommonElements(IHierarchy<T> other) {
        Set<T> result = new LinkedHashSet<>();
        Set<T> valuesOne = new LinkedHashSet<>();
        Set<T> valuesTwo = new LinkedHashSet<>();

        other.iterator().forEachRemaining(x -> {
            valuesOne.add(x);
        });

        this.iterator().forEachRemaining(x -> {
            valuesTwo.add(x);
        });

        valuesTwo.forEach(x -> {
            if (valuesOne.contains(x)) {
                result.add(x);
            }
        });

        return () -> result.iterator();
    }

    @Override
    public Iterator<T> iterator() {
        Set<T> result = new LinkedHashSet<>();

        Queue<Node<T>> nodes = new ArrayDeque<>();
        nodes.add(this.root);

        while (nodes.size() != 0) {
            Node node = nodes.peek();

            result.add((T) node.getValue());

            node.getChildrens().forEach(x -> {
                nodes.add((Node<T>) x);
            });

            nodes.poll();
        }

        return result.iterator();
    }
}

File: Data Structures/#11.B-Trees & Red-Black Trees Exercises/C#/Hierarchy.Core/Hierarchy.cs
namespace Hierarchy.Core
{
    using System;
    using System.Collections.Generic;
    using System.Collections;

    public class Hierarchy<T> : IHierarchy<T>
    {

        private Node<T> root;
        public Hierarchy(T root)
        {
            this.root = new Node<T>(root, null);
        }

        public int Count
        {
            get
            {
                return this.root.Count;
            }
        }

        public void Add(T element, T child)
        {
            Node<T> resultNode = null;
            bool isElementAlreadyPresent = false;
            Queue<Node<T>> nodes = new Queue<Node<T>>();
            nodes.Enqueue(this.root);

            while (nodes.Count != 0)
            {
                Node<T> current = nodes.Peek();

                if (current.Value.Equals(element))
                {
                    resultNode = current;
                }

                if (current.Value.Equals(child))
                {
                    isElementAlreadyPresent = true;
                    break;
                }

                current.Childrens.ForEach(x =>
                {
                    nodes.Enqueue(x);
                });

                nodes.Dequeue();
            }

            if (resultNode == null)
            {
                throw new ArgumentException("Parent doesn't exist!");
            }

            if (isElementAlreadyPresent)
            {
                throw new ArgumentException("Element already exists!");
            }

            resultNode.AddChildren(new Node<T>(child, resultNode));

        }

        public void Remove(T element)
        {
            if (this.root == null || this.root.Value.Equals(element))
            {
                throw new InvalidOperationException("You cannot delete the root node!");
            }

            Node<T> resultNode = null;
            Queue<Node<T>> nodes = new Queue<Node<T>>();
            nodes.Enqueue(this.root);

            while (nodes.Count != 0)
            {
                Node<T> current = nodes.Peek();

                if (current.Value.Equals(element))
                {
                    resultNode = current;
                    break;
                }

                current.Childrens.ForEach(x =>
                {
                    nodes.Enqueue(x);
                });

                nodes.Dequeue();
            }

            if (resultNode == null)
            {
                throw new ArgumentException("Element doesnt exists!");
            }

            Node<T> father = resultNode.Father;
            resultNode.UpdateFatherCountWithMinusOne();
            int resultNodeIndex = father.Childrens.IndexOf(resultNode);

            for (int i = 0; i < resultNode.Childrens.Count; i++)
            {
                father.Childrens.Insert(resultNodeIndex, resultNode.Childrens[i]);
                resultNodeIndex++;
            }

            father.Childrens.RemoveAt(resultNodeIndex);

        }

        public IEnumerable<T> GetChildren(T item)
        {
            Node<T> resultNode = null;
            Queue<Node<T>> nodes = new Queue<Node<T>>();
            nodes.Enqueue(this.root);

            while (nodes.Count != 0)
            {
                Node<T> current = nodes.Peek();

                if (current.Value.Equals(item))
                {
                    resultNode = current;
                    break;
                }

                current.Childrens.ForEach(x =>
                {
                    nodes.Enqueue(x);
                });

                nodes.Dequeue();
            }

            if (resultNode == null)
            {
                throw new ArgumentException("Element doesnt exists!");
            }

            List<T> childrens = new List<T>();

            resultNode.Childrens.ForEach(x =>
            {
                childrens.Add(x.Value);
            });

            return childrens;
        }

        public T GetParent(T item)
        {
            Node<T> resultNode = null;
            Queue<Node<T>> nodes = new Queue<Node<T>>();
            nodes.Enqueue(this.root);

            while (nodes.Count != 0)
            {
                Node<T> current = nodes.Peek();

                if (current.Value.Equals(item))
                {
                    resultNode = current;
                    break;
                }

                current.Childrens.ForEach(x =>
                {
                    nodes.Enqueue(x);
                });

                nodes.Dequeue();
            }

            if (resultNode == null)
            {
                throw new ArgumentException("Element doesnt exists!");
            }

            if (resultNode.Father != null)
            {
                return resultNode.Father.Value;
            }

            return default(T);
        }

        public bool Contains(T value)
        {
            bool isPresent = false;
            Queue<Node<T>> nodes = new Queue<Node<T>>();
            nodes.Enqueue(this.root);

            while (nodes.Count != 0)
            {
                Node<T> current = nodes.Peek();

                if (current.Value.Equals(value))
                {
                    isPresent = true;
                    break;
                }

                current.Childrens.ForEach(x =>
                {
                    nodes.Enqueue(x);
                });

                nodes.Dequeue();
            }

            return isPresent;
        }

        public IEnumerable<T> GetCommonElements(Hierarchy<T> other)
        {
            List<T> commonElements = new List<T>();

            foreach(var ele in other)
            {
                bool isPresent = this.Contains(ele);
                if(isPresent) commonElements.Add(ele);
            }

            return commonElements;
        }

        public IEnumerator<T> GetEnumerator()
        {
            List<T> result = new List<T>();
            Queue<Node<T>> nodes = new Queue<Node<T>>();
            nodes.Enqueue(this.root);

            while (nodes.Count != 0)
            {
                Node<T> current = nodes.Peek();

                result.Add(current.Value);
                current.Childrens.ForEach(x =>
                {
                    nodes.Enqueue(x);
                });

                nodes.Dequeue();
            }

            return result.GetEnumerator();
        }

        IEnumerator IEnumerable.GetEnumerator()
        {
            return this.GetEnumerator();
        }
    }
}